src/operators.jl

```diff
@@ -348,14 +348,14 @@ function Base.issymmetric{T<:JuMPTypes}(x::Matrix{T})
 end
 
 # Special-case because the the base version wants to do fill!(::Array{Variable}, zero(AffExpr))
-Base.diagm(x::Vector{Variable}) = diagm(convert(Vector{AffExpr}, x))
+Base.diagm(x::AbstractVector{Variable}) = diagm(convert(Vector{AffExpr}, x))
 
 ###############
 # The _multiply!(buf,y,z) adds the results of y*z into the buffer buf. No bounds/size
 # checks are performed; it is expected that the caller has done this, has ensured
 # that the eltype of buf is appropriate, and has zeroed the elements of buf (if desired).
 
-function _multiply!{T<:JuMPTypes}(ret::Array{T}, lhs::Array, rhs::Array)
+function _multiply!{T<:JuMPTypes}(ret::AbstractArray{T}, lhs::Array, rhs::Array)
```

Comment:
Yes, I don't think this part of the code will work with generic abstract arrays as-is.